cumulusci/utils/http/multi_request.py

```diff
@@ -100,13 +127,62 @@ def __enter__(self, *args):
     def __exit__(self, *args):
         self.close()
 
-    def do_composite_requests(self, requests):
+    def do_composite_requests(
+        self, requests
+    ) -> T.Tuple[T.Sequence, T.Sequence]:  # results, erros
         if not self.psf:
             raise AssertionError(
                 "Session was not opened. Please call open() or use as a context manager"
             )
 
         composite_requests = create_composite_requests(requests, self.chunk_size)
-        composite_results = self.psf.do_requests(composite_requests)
+        composite_results, errors = self.psf.do_requests(composite_requests)
         individual_results = parse_composite_results(composite_results)
-        return list(individual_results)
+
+        errors = list(errors)
+        if errors:
+            singleton_results, unrecoverable_errors = self.retry_errors(errors)
+        else:
+            singleton_results = []
+            unrecoverable_errors = []
+
+        individual_results = list(individual_results)
+
+        individual_results += list(singleton_results)
+        return individual_results, unrecoverable_errors
+
+    def retry_errors(self, errors):
+        unrecoverable_errors = []
+        singleton_requests = []
+        for exception, failed_composite_request in errors:
+            if isinstance(exception, RECOVERABLE_ERRORS):
+                singleton_requests.extend(split_requests(failed_composite_request))
+            else:
+                unrecoverable_errors.append((exception, failed_composite_request))
+
+        singleton_results, errors = self.psf.do_requests(singleton_requests)
+
+        def response_to_dict(response):
+            return {
+                "httpStatusCode": response.status_code,
+                "body": response.json(),
+                "httpHeaders": response.headers,
+            }
+
+        singleton_results = [
+            response_to_dict(response) for response in singleton_results
+        ]
+        errors = list(errors)
+
+        return singleton_results, unrecoverable_errors + errors
+
+
+def split_requests(composite_request: dict):
+    # need to remove this prefix-pattern because it will be added again laters
+    prefix = "/services/data/vxx.x"
+    single_requests = [r.copy() for r in composite_request["json"]["compositeRequest"]]
+    for request in single_requests:
+        del request["referenceId"]
+        if request["url"].startswith(prefix[0:-4]):
```

Comment:
Why would it literally start with vxx.x? Is this something you ended up with because of your integration testing's normalization that won't work for real?

Reply:
@prescod no, it doesn't. I don't understand why the composite API requests would have the literal string "vxx.x" instead of a specific version like "v52.0"

Reply:
The composite APIs do have a literal version string in them. The version pattern is used for strlen purposes.

Reply:
@prescod I get it now; I was misreading the range passed to startswith as including the version rather than excluding it.

Might be safer to split on / and work with indexed parts so that we don't suddenly break when the API gets to 3 digits.

Reply:
Fair enough; 9cd9d1613995810d085984db9a0622a0b25787de